feat(logs): user-selectable logs limit

MacDada opened the pull request:
When viewing debug logs, the current buffer limit of a 100 is full pretty quick.

I'm adding an option for the user to choose a bigger limit (up to a 1000).

<img width="1042" alt="Screenshot 2024-04-22 at 03 53 43" src="https://github.com/nurikk/zigbee2mqtt-frontend/assets/136261/661a2838-e713-4575-aa46-81ac22b260d3">

---

Please feel free to correct me, as I've never used React before ;)

MacDada committed:
feat(logs): user-selectable logs limit

When viewing debug logs, the current buffer limit of a 100 is full pretty quick.

I'm adding an option for the user to choose a bigger limit (up to a 1000).

MacDada commented on src/components/logs-page/index.tsx:
```diff
@@ -125,6 +126,28 @@ class LogsPage extends Component<PropsFromStore & BridgeApi & UtilsApi & WithTra
+                                }}
+                            >
+                                {logLimits.map((limit) => (
+                                    <option key={limit} value={limit} selected={limit == store.getState().logsLimit}>
```
i guess the `key` and `value` are unnecessary and can be removed?

i put them there, as i kinda copy-pasted `log-level` section ;)

nurikk-sa replied:
please refactor store.getState().logsLimit out of this loop, somewhere to the top of render function, example:

`const { logsLimit } = store.getState();`

MacDada replied:
BTW, when I make a commit responding to a comment, I like to link the comment in the commit description (so that it gives more opportunity to discover why a particular change was made).

How do I link to _this_ comment? ;-)

My commit message:

```
perf(logs): extracted var to optimize a loop

https://github.com/nurikk/zigbee2mqtt-frontend/pull/1999/commits/9bca1ad6688ad003b85faf0de4acb4c9e467e075#r1574572268
```

The result:

```
22:46:37.622: [zigbee2mqtt-frontend] git -c credential.helper= -c core.quotepath=false -c log.showSignature=false add --ignore-errors -A -f -- src/components/logs-page/index.tsx
22:46:37.630: [zigbee2mqtt-frontend] git -c credential.helper= -c core.quotepath=false -c log.showSignature=false commit -F /private/var/folders/ws/nzgf19n976gfpx235sryx8lc0000gn/T/git-commit-msg-.txt --
⧗   input: perf(logs): extracted var to optimize a loop
https://github.com/nurikk/zigbee2mqtt-frontend/pull/1999/commits/9bca1ad6688ad003b85faf0de4acb4c9e467e075#r1574572268
✖   footer's lines must not be longer than 100 characters [footer-max-line-length]
✖   found 1 problems, 0 warnings
ⓘ   Get help: https://github.com/conventional-changelog/commitlint/#what-is-commitlint
husky - commit-msg hook exited with code 1 (error)
```

MacDada replied:
done https://github.com/nurikk/zigbee2mqtt-frontend/pull/1999/commits/94b0f5005e8dae66a0ac5d923315dc16441f3189

nurikk-sa commented on src/components/logs-page/index.tsx:
```diff
@@ -125,6 +126,28 @@ class LogsPage extends Component<PropsFromStore & BridgeApi & UtilsApi & WithTra
+                                    const limit = parseInt(e.target.value);
+                                    store.setState({
+                                        logsLimit: limit,
+                                        logs: [...store.getState().logs.slice(-limit)],
```
I don't think we should slice logs here, this will happen automatically once new log record arrives, and then handler can be simplified to 
`store.setState({logsLimit: parseInt(e.target.value)})`

MacDada replied:
> this will happen automatically once new log record arrives

yep, but I think it's better for the user experience, if the slice is also done immediately upon selecting the option from the menu

it doesn't matter for 100 -> 1000 change, but it does in the case of changing 1000 -> 100 ;)

i don't think the extra slicing complicates this code so much, to make a worse UI instead…

Nerivec commented:
Sidenote:
The new unified logging (available in next Z2M release or currently in dev) will change the interaction with frontend a bit:
- `debug` level is no longer published to frontend/MQTT, only in console/file/syslog. _A note should probably be added in frontend's log page select to make this clear._
- "MQTT Publish" lines are now in `debug` level (also removing the need for frontend's blacklist for notifies).

More in below PRs:
https://github.com/Koenkk/zigbee2mqtt/pull/21984
https://github.com/Koenkk/zigbee-herdsman/pull/989
https://github.com/Koenkk/zigbee2mqtt/pull/22066
https://github.com/Koenkk/zigbee2mqtt/pull/22129

MacDada committed:
perf(logs): extracted var to optimize a loop

https://github.com/nurikk/zigbee2mqtt-frontend/pull/1999/commits/
9bca1ad6688ad003b85faf0de4acb4c9e467e075#r1574572268
